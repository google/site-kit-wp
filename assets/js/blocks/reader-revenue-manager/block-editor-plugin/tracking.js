import { trackEvent } from '@/js/util';

export function trackBlockUsage(
	blockName,
	action,
	label = null,
	trackGAEvent = trackEvent
) {
	trackGAEvent( blockName, action, label );
}
