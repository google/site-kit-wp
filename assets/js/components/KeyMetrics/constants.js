/**
 * Key Metrics components - constants.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

export const KEY_METRICS_SETUP_CTA_WIDGET_SLUG = 'key-metrics-setup-cta-widget';

export const KEY_METRICS_SELECTION_PANEL_OPENED_KEY =
	'googlesitekit-key-metrics-selection-panel-opened';
export const KEY_METRICS_SELECTION_FORM = 'key-metrics-selection-form';
export const KEY_METRICS_SELECTED = 'key-metrics-selected';
export const EFFECTIVE_SELECTION = 'key-metrics-effective-selection';
export const UNSTAGED_SELECTION = 'key-metrics-unstaged-selection';
export const MIN_SELECTED_METRICS_COUNT = 2;
export const MAX_SELECTED_METRICS_COUNT = 4;
export const MAX_SELECTED_METRICS_COUNT_WITH_CONVERSION_EVENTS = 8;
export const KEY_METRICS_GROUP_CURRENT = {
	SLUG: 'current-selection',
	LABEL: __( 'Current selection', 'google-site-kit' ),
};
export const KEY_METRICS_GROUP_SUGGESTED = {
	SLUG: 'suggested',
	LABEL: __( 'Suggested', 'google-site-kit' ),
};
export const KEY_METRICS_GROUP_VISITORS = {
	SLUG: 'visitors',
	LABEL: __( 'Visitors', 'google-site-kit' ),
};
export const KEY_METRICS_GROUP_DRIVING_TRAFFIC = {
	SLUG: 'driving-traffic',
	LABEL: __( 'Driving traffic', 'google-site-kit' ),
};
export const KEY_METRICS_GROUP_GENERATING_LEADS = {
	SLUG: 'generating-leads',
	LABEL: __( 'Generating leads', 'google-site-kit' ),
};
export const KEY_METRICS_GROUP_SELLING_PRODUCTS = {
	SLUG: 'selling-products',
	LABEL: __( 'Selling products', 'google-site-kit' ),
};
export const KEY_METRICS_GROUP_CONTENT_PERFORMANCE = {
	SLUG: 'content-performance',
	LABEL: __( 'Content performance', 'google-site-kit' ),
};
