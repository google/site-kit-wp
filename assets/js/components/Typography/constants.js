/**
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const TYPE_DISPLAY = 'display';
export const TYPE_HEADLINE = 'headline';
export const TYPE_TITLE = 'title';
export const TYPE_BODY = 'body';
export const TYPE_LABEL = 'label';

export const VALID_TYPES = [
	TYPE_BODY,
	TYPE_DISPLAY,
	TYPE_HEADLINE,
	TYPE_LABEL,
	TYPE_TITLE,
];

export const SIZE_SMALL = 'small';
export const SIZE_MEDIUM = 'medium';
export const SIZE_LARGE = 'large';

export const VALID_SIZES = [ SIZE_SMALL, SIZE_MEDIUM, SIZE_LARGE ];
