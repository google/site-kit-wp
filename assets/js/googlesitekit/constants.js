/**
 * Core constants.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const VIEW_CONTEXT_MAIN_DASHBOARD = 'mainDashboard';
export const VIEW_CONTEXT_ENTITY_DASHBOARD = 'entityDashboard';
export const VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY = 'mainDashboardViewOnly';
export const VIEW_CONTEXT_ENTITY_DASHBOARD_VIEW_ONLY =
	'entityDashboardViewOnly';
export const VIEW_CONTEXT_POSTS_LIST = 'postsList';
export const VIEW_CONTEXT_USER_INPUT = 'userInput';
export const VIEW_CONTEXT_ACTIVATION = 'activation';
export const VIEW_CONTEXT_SPLASH = 'splash';
export const VIEW_CONTEXT_ADMIN_BAR = 'adminBar';
export const VIEW_CONTEXT_ADMIN_BAR_VIEW_ONLY = 'adminBarViewOnly';
export const VIEW_CONTEXT_SETTINGS = 'settings';
export const VIEW_CONTEXT_AD_BLOCKING_RECOVERY = 'adBlockingRecovery';
export const VIEW_CONTEXT_MODULE = 'module';
export const VIEW_CONTEXT_WP_DASHBOARD = 'wpDashboard';
export const VIEW_CONTEXT_WP_DASHBOARD_VIEW_ONLY = 'wpDashboardViewOnly';
export const VIEW_CONTEXT_MODULE_SETUP = 'moduleSetup';
export const VIEW_CONTEXT_METRIC_SELECTION = 'metricSelection';

// Anchor ID Constants
export const ANCHOR_ID_KEY_METRICS = 'key-metrics';
export const ANCHOR_ID_TRAFFIC = 'traffic';
export const ANCHOR_ID_CONTENT = 'content';
export const ANCHOR_ID_SPEED = 'speed';
export const ANCHOR_ID_MONETIZATION = 'monetization';

// Site Kit only view contexts
export const SITE_KIT_VIEW_CONTEXTS = [
	VIEW_CONTEXT_MAIN_DASHBOARD,
	VIEW_CONTEXT_ENTITY_DASHBOARD,
	VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY,
	VIEW_CONTEXT_ENTITY_DASHBOARD_VIEW_ONLY,
	VIEW_CONTEXT_USER_INPUT,
	VIEW_CONTEXT_SPLASH,
	VIEW_CONTEXT_SETTINGS,
	VIEW_CONTEXT_MODULE_SETUP,
	VIEW_CONTEXT_METRIC_SELECTION,
];

// Site Kit view-only contexts
export const SITE_KIT_VIEW_ONLY_CONTEXTS = [
	VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY,
	VIEW_CONTEXT_ENTITY_DASHBOARD_VIEW_ONLY,
	VIEW_CONTEXT_ADMIN_BAR_VIEW_ONLY,
	VIEW_CONTEXT_WP_DASHBOARD_VIEW_ONLY,
];
