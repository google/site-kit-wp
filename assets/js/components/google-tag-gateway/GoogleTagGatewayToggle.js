/**
 * Google Tag Gateway Toggle component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import classnames from 'classnames';
import { useMount } from 'react-use';

/**
 * WordPress dependencies
 */
import { createInterpolateElement, useCallback } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { ProgressBar, Switch } from 'googlesitekit-components';
import { useDispatch, useSelect } from 'googlesitekit-data';
import { CORE_SITE } from '@/js/googlesitekit/datastore/site/constants';
import NewBadge from '@/js/components/NewBadge';
import Link from '@/js/components/Link';
import useViewContext from '@/js/hooks/useViewContext';
import { trackEvent } from '@/js/util';
import withIntersectionObserver from '@/js/util/withIntersectionObserver';
import Notice from '@/js/components/Notice';
import GoogleTagGatewayOptOutNotice from './GoogleTagGatewayOptOutNotice';

const SubtleNotificationWithIntersectionObserver =
	withIntersectionObserver( Notice );

export default function GoogleTagGatewayToggle( { className } ) {
	const viewContext = useViewContext();

	const isGoogleTagGatewayEnabled = useSelect( ( select ) =>
		select( CORE_SITE ).isGoogleTagGatewayEnabled()
	);
	const isLoading = useSelect( ( select ) =>
		select( CORE_SITE ).isFetchingGetGTGServerRequirementStatus()
	);
	const hasMetServerRequirements = useSelect( ( select ) => {
		const { isGTGHealthy, isScriptAccessEnabled } = select( CORE_SITE );

		return isGTGHealthy() !== false && isScriptAccessEnabled() !== false;
	} );

	const { fetchGetGTGServerRequirementStatus, setGoogleTagGatewayEnabled } =
		useDispatch( CORE_SITE );

	const learnMoreURL = useSelect( ( select ) => {
		return select( CORE_SITE ).getDocumentationLinkURL(
			'google-tag-gateway-introduction'
		);
	} );

	const serverRequirementsLearnMoreURL = useSelect( ( select ) => {
		return select( CORE_SITE ).getDocumentationLinkURL(
			'google-tag-gateway-server-requirements'
		);
	} );

	// Fetch the server requirement status on mount.
	useMount( fetchGetGTGServerRequirementStatus );

	const handleClick = useCallback( () => {
		const action = isGoogleTagGatewayEnabled
			? 'deactivate_google_tag_gateway'
			: 'activate_google_tag_gateway';

		trackEvent( `${ viewContext }_gtg-settings-toggle`, action ).finally(
			() => {
				setGoogleTagGatewayEnabled( ! isGoogleTagGatewayEnabled );
			}
		);
	}, [ isGoogleTagGatewayEnabled, setGoogleTagGatewayEnabled, viewContext ] );

	return (
		<div
			className={ classnames(
				'googlesitekit-google-tag-gateway-toggle',
				className
			) }
		>
			{ isLoading && (
				<ProgressBar
					className="googlesitekit-google-tag-gateway-toggle__progress"
					small
				/>
			) }
			{ ! isLoading && (
				<div className="googlesitekit-module-settings-group__switch">
					<Switch
						label={ __(
							'Google tag gateway for advertisers',
							'google-site-kit'
						) }
						checked={
							!! isGoogleTagGatewayEnabled &&
							hasMetServerRequirements
						}
						disabled={ ! hasMetServerRequirements }
						onClick={ handleClick }
						hideLabel={ false }
					/>
					<div className="googlesitekit-google-tag-gateway-toggle__switch-badge">
						<NewBadge hasLeftSpacing />
					</div>
				</div>
			) }
			<p className="googlesitekit-module-settings-group__helper-text">
				{ createInterpolateElement(
					__(
						'Your tag data will be sent through your own domain to improve data quality and help you recover measurement signals. <a>Learn more</a>',
						'google-site-kit'
					),
					{
						a: (
							<Link
								href={ learnMoreURL }
								onClick={ () => {
									trackEvent(
										`${ viewContext }_gtg-settings-toggle`,
										'click_learn_more_link'
									);
								} }
								aria-label={ __(
									'Learn more about Google tag gateway for advertisers',
									'google-site-kit'
								) }
								external
							/>
						),
					}
				) }
			</p>
			<GoogleTagGatewayOptOutNotice />
			{ ! isLoading && ! hasMetServerRequirements && (
				<SubtleNotificationWithIntersectionObserver
					type={ Notice.TYPES.WARNING }
					title={ createInterpolateElement(
						__(
							'Your serverâ€™s current settings prevent Google tag gateway for advertisers from working. To enable it, please contact your hosting provider and request access to external resources and plugin files. <a>Learn more</a>',
							'google-site-kit'
						),
						{
							a: (
								<Link
									href={ serverRequirementsLearnMoreURL }
									onClick={ () => {
										trackEvent(
											`${ viewContext }_gtg-settings-toggle-disabled`,
											'click_learn_more_link'
										);
									} }
									aria-label={ __(
										'Learn more about Google tag gateway for advertisers server requirements',
										'google-site-kit'
									) }
									external
								/>
							),
						}
					) }
					variant="warning"
					onInView={ () => {
						trackEvent(
							`${ viewContext }_gtg-settings-toggle-disabled`,
							'view_notice'
						);
					} }
				/>
			) }
		</div>
	);
}
