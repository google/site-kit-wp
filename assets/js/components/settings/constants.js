/**
 * Settings constants.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { MODULE_SLUG_ADS } from '../../modules/ads/datastore/constants';
import { MODULE_SLUG_SIGN_IN_WITH_GOOGLE } from '../../modules/sign-in-with-google/datastore/constants';
import { MODULE_SLUG_READER_REVENUE_MANAGER } from '../../modules/reader-revenue-manager/datastore/constants';

export const NEW_MODULES = [
	MODULE_SLUG_ADS,
	MODULE_SLUG_READER_REVENUE_MANAGER,
	MODULE_SLUG_SIGN_IN_WITH_GOOGLE,
];

export const BETA_MODULES = [ MODULE_SLUG_SIGN_IN_WITH_GOOGLE ];

export const EXPERIMENTAL_MODULES = [];
