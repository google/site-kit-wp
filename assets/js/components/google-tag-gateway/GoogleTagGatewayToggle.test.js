/**
 * Google Tag Gateway Toggle component tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { useIntersection as mockUseIntersection } from 'react-use';

/**
 * Internal dependencies
 */
import {
	render,
	createTestRegistry,
	muteFetch,
	freezeFetch,
	act,
	waitForDefaultTimeouts,
} from '../../../../tests/js/test-utils';
import { CORE_SITE } from '@/js/googlesitekit/datastore/site/constants';
import { VIEW_CONTEXT_MAIN_DASHBOARD } from '@/js/googlesitekit/constants';
import * as tracking from '@/js/util/tracking';
import GoogleTagGatewayToggle from './GoogleTagGatewayToggle';

jest.mock( 'react-use', () => ( {
	...jest.requireActual( 'react-use' ),
	useIntersection: jest.fn(),
} ) );

const mockTrackEvent = jest.spyOn( tracking, 'trackEvent' );
mockTrackEvent.mockImplementation( () => Promise.resolve() );

describe( 'GoogleTagGatewayToggle', () => {
	let registry;

	const serverRequirementStatusEndpoint = new RegExp(
		'^/google-site-kit/v1/core/site/data/gtg-server-requirement-status'
	);

	beforeEach( () => {
		mockUseIntersection.mockImplementation( () => ( {
			isIntersecting: false,
			intersectionRatio: 0,
		} ) );

		registry = createTestRegistry();

		registry.dispatch( CORE_SITE ).receiveGetGoogleTagGatewaySettings( {
			isEnabled: false,
			isGTGHealthy: null,
			isScriptAccessEnabled: null,
		} );
	} );

	afterEach( () => {
		jest.clearAllMocks();
	} );

	it( 'should make a request to fetch the server requirement status', async () => {
		muteFetch( serverRequirementStatusEndpoint );

		const { waitForRegistry } = render( <GoogleTagGatewayToggle />, {
			registry,
		} );

		await waitForRegistry();

		expect( fetchMock ).toHaveFetched( serverRequirementStatusEndpoint );
	} );

	it( 'should render in loading state if the server requirement status is still loading', async () => {
		freezeFetch( serverRequirementStatusEndpoint );

		const { container, getByRole, waitForRegistry } = render(
			<GoogleTagGatewayToggle />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( getByRole( 'progressbar' ) ).toBeInTheDocument();

		expect( container ).toMatchSnapshot();
	} );

	it( 'should render in default state', async () => {
		fetchMock.getOnce( serverRequirementStatusEndpoint, {
			body: {
				isEnabled: false,
				isGTGHealthy: true,
				isScriptAccessEnabled: true,
			},
			status: 200,
		} );

		const { container, getByLabelText, waitForRegistry } = render(
			<GoogleTagGatewayToggle />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect(
			getByLabelText( 'Google tag gateway for advertisers' )
		).toBeInTheDocument();

		expect( container ).toMatchSnapshot();
	} );

	it( 'should render in disabled state if server requirements are not met', async () => {
		fetchMock.getOnce( serverRequirementStatusEndpoint, {
			body: {
				isEnabled: false,
				isGTGHealthy: false,
				isScriptAccessEnabled: false,
			},
			status: 200,
		} );

		const { container, getByLabelText, waitForRegistry } = render(
			<GoogleTagGatewayToggle />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect(
			getByLabelText( 'Google tag gateway for advertisers' )
		).toBeDisabled();

		expect( container ).toHaveTextContent(
			'Your server’s current settings prevent Google tag gateway for advertisers from working. To enable it, please contact your hosting provider and request access to external resources and plugin files.'
		);

		expect( container ).toMatchSnapshot();
	} );

	it( 'should not track an event when the toggle is viewed with no warning notice', async () => {
		fetchMock.getOnce( serverRequirementStatusEndpoint, {
			body: {
				isEnabled: false,
				isGTGHealthy: true,
				isScriptAccessEnabled: true,
			},
			status: 200,
		} );

		const { getByLabelText, queryByText, rerender, waitForRegistry } =
			render( <GoogleTagGatewayToggle />, {
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			} );

		await waitForRegistry();

		expect(
			getByLabelText( 'Google tag gateway for advertisers' )
		).toBeInTheDocument();

		expect(
			queryByText(
				'Your server’s current settings prevent Google tag gateway for advertisers from working. To enable it, please contact your hosting provider and request access to external resources and plugin files.'
			)
		).not.toBeInTheDocument();

		// Simulate the warning notice becoming visible if it were present.
		mockUseIntersection.mockImplementation( () => ( {
			isIntersecting: true,
			intersectionRatio: 1,
		} ) );

		rerender( <GoogleTagGatewayToggle /> );

		expect( mockTrackEvent ).not.toHaveBeenCalled();
	} );

	it( 'should track an event when the warning notice is viewed', async () => {
		fetchMock.getOnce( serverRequirementStatusEndpoint, {
			body: {
				isEnabled: false,
				isGTGHealthy: false,
				isScriptAccessEnabled: false,
			},
			status: 200,
		} );

		const { getByLabelText, getByText, rerender, waitForRegistry } = render(
			<GoogleTagGatewayToggle />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		await waitForRegistry();

		expect(
			getByLabelText( 'Google tag gateway for advertisers' )
		).toBeInTheDocument();

		expect(
			getByText(
				'Your server’s current settings prevent Google tag gateway for advertisers from working. To enable it, please contact your hosting provider and request access to external resources and plugin files.'
			)
		).toBeInTheDocument();

		mockUseIntersection.mockImplementation( () => ( {
			isIntersecting: true,
			intersectionRatio: 1,
		} ) );

		// Simulate the warning notice becoming visible.
		rerender( <GoogleTagGatewayToggle /> );

		expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
		expect( mockTrackEvent ).toHaveBeenCalledWith(
			'mainDashboard_gtg-settings-toggle-disabled',
			'view_notice'
		);
	} );

	it( 'should track an event when the `Learn more` link is clicked in the toggle description', async () => {
		fetchMock.getOnce( serverRequirementStatusEndpoint, {
			body: {
				isEnabled: false,
				isGTGHealthy: true,
				isScriptAccessEnabled: true,
			},
			status: 200,
		} );

		const { getByRole, waitForRegistry } = render(
			<GoogleTagGatewayToggle />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		await waitForRegistry();

		expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

		const learnMoreLink = getByRole( 'link', {
			name: 'Learn more about Google tag gateway for advertisers (opens in a new tab)',
		} );

		learnMoreLink.click();

		expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
		expect( mockTrackEvent ).toHaveBeenCalledWith(
			'mainDashboard_gtg-settings-toggle',
			'click_learn_more_link'
		);
	} );

	it( 'should track an event when the `Learn more` link is clicked in the warning notice', async () => {
		fetchMock.getOnce( serverRequirementStatusEndpoint, {
			body: {
				isEnabled: false,
				isGTGHealthy: false,
				isScriptAccessEnabled: false,
			},
			status: 200,
		} );

		const { getByRole, waitForRegistry } = render(
			<GoogleTagGatewayToggle />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		await waitForRegistry();

		expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

		const learnMoreLink = getByRole( 'link', {
			name: 'Learn more about Google tag gateway for advertisers server requirements (opens in a new tab)',
		} );

		learnMoreLink.click();

		expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
		expect( mockTrackEvent ).toHaveBeenCalledWith(
			'mainDashboard_gtg-settings-toggle-disabled',
			'click_learn_more_link'
		);
	} );

	it.each( [ 'isGTGHealthy', 'isScriptAccessEnabled' ] )(
		'should not render in disabled state unless %s is explicitly false',
		async ( requirement ) => {
			const response = {
				isEnabled: false,
				isGTGHealthy: true,
				isScriptAccessEnabled: true,
			};

			response[ requirement ] = null;

			fetchMock.getOnce( serverRequirementStatusEndpoint, {
				body: response,
				status: 200,
			} );

			const { getByLabelText, waitForRegistry } = render(
				<GoogleTagGatewayToggle />,
				{
					registry,
				}
			);

			await waitForRegistry();

			expect(
				getByLabelText( 'Google tag gateway for advertisers' )
			).not.toBeDisabled();
		}
	);

	it.each( [ 'isGTGHealthy', 'isScriptAccessEnabled' ] )(
		'should render in disabled state if %s is false',
		async ( requirement ) => {
			const response = {
				isEnabled: false,
				isGTGHealthy: true,
				isScriptAccessEnabled: true,
			};

			response[ requirement ] = false;

			fetchMock.getOnce( serverRequirementStatusEndpoint, {
				body: response,
				status: 200,
			} );

			const { getByLabelText, waitForRegistry } = render(
				<GoogleTagGatewayToggle />,
				{
					registry,
				}
			);

			await waitForRegistry();

			expect(
				getByLabelText( 'Google tag gateway for advertisers' )
			).toBeDisabled();
		}
	);

	it( 'should toggle Google tag gateway for advertisers on click', async () => {
		fetchMock.getOnce( serverRequirementStatusEndpoint, {
			body: {
				isEnabled: false,
				isGTGHealthy: true,
				isScriptAccessEnabled: true,
			},
			status: 200,
		} );

		const { getByLabelText, waitForRegistry } = render(
			<GoogleTagGatewayToggle />,
			{
				registry,
			}
		);

		await waitForRegistry();

		const switchControl = getByLabelText(
			'Google tag gateway for advertisers'
		);

		expect( switchControl ).not.toBeChecked();

		expect( registry.select( CORE_SITE ).isGoogleTagGatewayEnabled() ).toBe(
			false
		);

		switchControl.click();

		// Allow the `trackEvent()` promise to resolve.
		await act( waitForDefaultTimeouts );

		expect( switchControl ).toBeChecked();

		expect( registry.select( CORE_SITE ).isGoogleTagGatewayEnabled() ).toBe(
			true
		);

		// Give it another click to verify it can be toggled off.
		switchControl.click();

		// Allow the `trackEvent()` promise to resolve.
		await act( waitForDefaultTimeouts );

		expect( switchControl ).not.toBeChecked();

		expect( registry.select( CORE_SITE ).isGoogleTagGatewayEnabled() ).toBe(
			false
		);
	} );

	it( 'should track an event when the toggle is clicked', async () => {
		fetchMock.getOnce( serverRequirementStatusEndpoint, {
			body: {
				isEnabled: false,
				isGTGHealthy: true,
				isScriptAccessEnabled: true,
			},
			status: 200,
		} );

		const { getByLabelText, waitForRegistry } = render(
			<GoogleTagGatewayToggle />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		await waitForRegistry();

		const switchControl = getByLabelText(
			'Google tag gateway for advertisers'
		);

		switchControl.click();

		// Allow the `trackEvent()` promise to resolve.
		await act( waitForDefaultTimeouts );

		expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
		expect( mockTrackEvent ).toHaveBeenCalledWith(
			'mainDashboard_gtg-settings-toggle',
			'activate_google_tag_gateway'
		);

		switchControl.click();

		// Allow the `trackEvent()` promise to resolve.
		await act( waitForDefaultTimeouts );

		expect( mockTrackEvent ).toHaveBeenCalledTimes( 2 );
		expect( mockTrackEvent ).toHaveBeenCalledWith(
			'mainDashboard_gtg-settings-toggle',
			'deactivate_google_tag_gateway'
		);
	} );
} );
