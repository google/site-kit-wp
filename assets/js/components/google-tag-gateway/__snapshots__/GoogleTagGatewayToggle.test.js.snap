// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleTagGatewayToggle should render in default state 1`] = `
<div>
  <div
    class="googlesitekit-google-tag-gateway-toggle"
  >
    <div
      class="googlesitekit-module-settings-group__switch"
    >
      <div
        aria-checked="false"
        class="mdc-switch"
        role="switch"
        tabindex="0"
      >
        <div
          class="mdc-switch__track"
        >
           
        </div>
        <div
          class="mdc-switch__thumb-underlay"
        >
          <div
            class="mdc-switch__thumb"
          >
            <input
              class="mdc-switch__native-control"
              id="googlesitekit-switch-4"
              readonly=""
              role="switch"
              type="checkbox"
            />
          </div>
        </div>
      </div>
      <label
        class=""
        for="googlesitekit-switch-4"
      >
        Google tag gateway for advertisers
      </label>
      <div
        class="googlesitekit-google-tag-gateway-toggle__switch-badge"
      >
        <span
          class="googlesitekit-badge googlesitekit-new-badge googlesitekit-badge--has-left-spacing"
        >
          New
        </span>
      </div>
    </div>
    <p
      class="googlesitekit-module-settings-group__helper-text"
    >
      Your tag data will be sent through your own domain to improve data quality and help you recover measurement signals. 
      <a
        aria-label="Learn more about Google tag gateway for advertisers (opens in a new tab)"
        class="googlesitekit-cta-link"
        href="https://sitekit.withgoogle.com/support/?doc=google-tag-gateway-introduction"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          class="googlesitekit-cta-link__contents"
        >
          Learn more
        </span>
        <span
          class="googlesitekit-icon-wrapper"
          style="margin-left: 5px;"
        >
          <svg />
        </span>
      </a>
    </p>
  </div>
</div>
`;

exports[`GoogleTagGatewayToggle should render in disabled state if server requirements are not met 1`] = `
<div>
  <div
    class="googlesitekit-google-tag-gateway-toggle"
  >
    <div
      class="googlesitekit-module-settings-group__switch"
    >
      <div
        aria-checked="false"
        class="mdc-switch mdc-switch--disabled"
        role="switch"
        tabindex="0"
      >
        <div
          class="mdc-switch__track"
        >
           
        </div>
        <div
          class="mdc-switch__thumb-underlay"
        >
          <div
            class="mdc-switch__thumb"
          >
            <input
              class="mdc-switch__native-control"
              disabled=""
              id="googlesitekit-switch-6"
              readonly=""
              role="switch"
              type="checkbox"
            />
          </div>
        </div>
      </div>
      <label
        class=""
        for="googlesitekit-switch-6"
      >
        Google tag gateway for advertisers
      </label>
      <div
        class="googlesitekit-google-tag-gateway-toggle__switch-badge"
      >
        <span
          class="googlesitekit-badge googlesitekit-new-badge googlesitekit-badge--has-left-spacing"
        >
          New
        </span>
      </div>
    </div>
    <p
      class="googlesitekit-module-settings-group__helper-text"
    >
      Your tag data will be sent through your own domain to improve data quality and help you recover measurement signals. 
      <a
        aria-label="Learn more about Google tag gateway for advertisers (opens in a new tab)"
        class="googlesitekit-cta-link"
        href="https://sitekit.withgoogle.com/support/?doc=google-tag-gateway-introduction"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          class="googlesitekit-cta-link__contents"
        >
          Learn more
        </span>
        <span
          class="googlesitekit-icon-wrapper"
          style="margin-left: 5px;"
        >
          <svg />
        </span>
      </a>
    </p>
    <div
      class="googlesitekit-notice googlesitekit-notice--warning"
    >
      <div
        class="googlesitekit-notice__icon"
      >
        <svg />
      </div>
      <div
        class="googlesitekit-notice__content"
      >
        <p
          class="googlesitekit-notice__title"
        >
          Your server’s current settings prevent Google tag gateway for advertisers from working. To enable it, please contact your hosting provider and request access to external resources and plugin files. 
          <a
            aria-label="Learn more about Google tag gateway for advertisers server requirements (opens in a new tab)"
            class="googlesitekit-cta-link"
            href="https://sitekit.withgoogle.com/support/?doc=google-tag-gateway-server-requirements"
            rel="noopener noreferrer"
            target="_blank"
          >
            <span
              class="googlesitekit-cta-link__contents"
            >
              Learn more
            </span>
            <span
              class="googlesitekit-icon-wrapper"
              style="margin-left: 5px;"
            >
              <svg />
            </span>
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`GoogleTagGatewayToggle should render in loading state if the server requirement status is still loading 1`] = `
<div>
  <div
    class="googlesitekit-google-tag-gateway-toggle"
  >
    <div
      class="mdc-linear-progress googlesitekit-google-tag-gateway-toggle__progress mdc-linear-progress--indeterminate mdc-linear-progress--small"
      role="progressbar"
      style="height: 4px;"
    >
      <div
        class="mdc-linear-progress__buffering-dots"
      />
      <div
        class="mdc-linear-progress__buffer"
      />
      <div
        class="mdc-linear-progress__bar mdc-linear-progress__primary-bar"
      >
        <span
          class="mdc-linear-progress__bar-inner"
        />
      </div>
      <div
        class="mdc-linear-progress__bar mdc-linear-progress__secondary-bar"
      >
        <span
          class="mdc-linear-progress__bar-inner"
        />
      </div>
    </div>
    <p
      class="googlesitekit-module-settings-group__helper-text"
    >
      Your tag data will be sent through your own domain to improve data quality and help you recover measurement signals. 
      <a
        aria-label="Learn more about Google tag gateway for advertisers (opens in a new tab)"
        class="googlesitekit-cta-link"
        href="https://sitekit.withgoogle.com/support/?doc=google-tag-gateway-introduction"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          class="googlesitekit-cta-link__contents"
        >
          Learn more
        </span>
        <span
          class="googlesitekit-icon-wrapper"
          style="margin-left: 5px;"
        >
          <svg />
        </span>
      </a>
    </p>
  </div>
</div>
`;
